### DISCUSSION TITLE
What's the backend of the contour algorithms?

### DISCUSSION THREAD

**Discussion by dichen9412** [pyvista/pyvista#2572] [Q&A] [answered]
Hi there,

I think the "flying_edges" is the fastest contour algorithm in Pyvista (Correct me if I'm wrong.). May I know what's the back-end implementation of it?
is it done via high-speed language like c++/fortran? Or it is implemented in python?
I'm doing some heavy computation to generate a large mesh from grid data. I'm just wondering whether I can make that process faster if I use some C++ libraries. 

Best,
Di

**Comment by tkoyama010** [comment, marked as answer]
PyVista is a wrapper for vtk and the "flying_edges" option is also implemented as a wrapper for the [vtkFlyingEdges3D](https://vtk.org/doc/nightly/html/classvtkFlyingEdges3D.html) class of vtk. 
https://github.com/pyvista/pyvista/blob/c49a6abae7cc62d242f12ec45a6b22b524db1ec8/pyvista/core/filters/data_set.py#L1541-L1542
PyVista uses the vtk Python interface, but the engine part is [implemented in C++](https://vtk.org/doc/nightly/html/classvtkFlyingEdges3D.html).

**Reply by dichen9412** [reply]
Thanks! Then I guess I can't really find a C++ library that is significantly faster than pyvista.

**Reply by akaszynski** [reply]
From time to time I've found libraries that have superior performance to vtk's algorithms (e.g. [Fast-Quadric-Mesh-Simplification](https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification) versus VTK's [QuadricDecimation](https://vtk.org/doc/nightly/html/classvtkQuadricDecimation.html)). These sometimes come at a cost in terms or reliability, or sometimes they're remarkably performant (as in the case of [pygeodesic](https://github.com/pyvista/pyvista/discussions/2424)).

The main advantage of VTK and PyVista is to provide a convenient software library that's well documented, relatively performant, and contains several algorithms that are all exposed in a single library.

If you're looking for something faster, play around with https://github.com/nihaljn/marching-cubes and evaluate its performance.

**Reply by dichen9412** [reply]
Thanks for your suggestions! 
I'm mainly looking for a fast algorithm that can generate meshes from a binary grid mask. 
Thus, I'm more interested in a fast marching cube algorithm than a mesh simplification algorithm. I will try your suggestions for sure!

**Reply by darikg** [reply]
check out [pygalmesh](https://github.com/meshpro/pygalmesh/) maybe

**Reply by dichen9412** [reply]
Thanks, I will check it out!

**Reply by dichen9412** [reply]
Hi there, 

After profile the running time of my code, I find that actually, the most time-consuming operation is pv.save_meshio(<filename>, mesh). The mesh itself is computed really fast. 
Here is a snapshot of my code. 
`      grid = pv.UniformGrid()
        grid.dimensions = np.array(tmp.shape)
        grid.origin = (origin + spacing * (l - 1))
        grid.spacing = (spacing)
        grid.point_data["values"] = values
        print("step-2", time.time() - st)
        mesh = grid.contour(1, method="flying_edges", rng=[0.5, 0.5])
        print("step-3", time.time() - st)
        if (smooth_fac > 0):
            mesh = mesh.smooth(20, relaxation_factor=smooth_fac * 0.5)
        print("step-4", time.time() - st)

        filename = "mesh_%d.obj"%(i)
        pv.save_meshio("/tmp/" + filename, mesh)
`
Does anyone know if there is a better/faster way for saving the mesh? 

Best,
Di

**Reply by dichen9412** [reply]
Just to inform others, the function "mesh.save()" is much faster. The .ply format is as good as .obj format.
